### ISSUE TITLE
'dd' failure at launch of fstransform

### ISSUE THREAD

**Issue by rswarthout** [cosmos72/fstransform#47]
I am looking to convert a volume that is ~32TB in allocated size. Currently, the volume is 75% utilized. The attempted conversion is from `ext4` to `xfs` on a Rocky Linux 8 machine. Any guidance on how to proceed with be helpful.

```
$ sudo fstransform /dev/sdd xfs
fstransform: starting version 0.9.3, checking environment
fstransform: checking for which... 	'/bin/which'
fstransform: checking for expr... 	'/bin/expr'
fstransform: checking for id... 	'/bin/id'
fstransform: parsing command line arguments
fstransform: checking for stat... 	'/bin/stat'
fstransform: checking for mkfifo... 	'/bin/mkfifo'
fstransform: checking for blockdev... 	'/sbin/blockdev'
fstransform: checking for losetup... 	'/sbin/losetup'
fstransform: checking for fsck... 	'/sbin/fsck'
fstransform: checking for mkfs... 	'/sbin/mkfs'
fstransform: checking for mount... 	'/bin/mount'
fstransform: checking for umount... 	'/bin/umount'
fstransform: checking for mkdir... 	'/bin/mkdir'
fstransform: checking for rmdir... 	'/bin/rmdir'
fstransform: checking for rm... 	'/bin/rm'
fstransform: checking for dd... 	'/bin/dd'
fstransform: checking for sync... 	'/bin/sync'
fstransform: checking for fsmove... 	'/sbin/fsmove'
fstransform: checking for fsremap... 	'/sbin/fsremap'
fstransform: checking for fsck(source file-system)...	'/sbin/fsck'
fstransform: checking for fsck(target file-system)...	'/sbin/fsck'
fstransform: looking for optional commands
fstransform: checking for sleep... 	'/bin/sleep'
fstransform: checking for date... 	'/bin/date'
17:45:36 fstransform: environment check passed.
17:45:36 fstransform: saving output of this execution into /var/tmp/fstransform/fstransform.log.308145
17:45:36 fstransform: preparing to transform device '/dev/sdd' to file-system type 'xfs'
17:45:36 fstransform: device is mounted at '/virtual_0' with file-system type 'ext4'
17:45:36 fstransform: device raw size = 32635174780928 bytes
17:45:36 fstransform: creating sparse loop file '/virtual_0/.fstransform.loop.308145' inside device '/dev/sdd'...
17:45:36 dd: 1+0 records in
17:45:36 dd: 1+0 records out
17:45:36 dd: 1 byte copied, 5.2134e-05 s, 19.2 kB/s
17:45:36 fstransform: device file-system block size = 4096 bytes
17:45:36 fstransform: device usable size = 32635174780928 bytes
17:45:36 dd: failed to truncate to 32635174780927 bytes in output file '/virtual_0/.fstransform.loop.308145': File too large
         
17:45:36 ERROR! fstransform: command '/bin/dd if=/dev/zero of=/virtual_0/.fstransform.loop.308145 bs=1 count=1 seek=32635174780927' failed (exit status 1)
                this is potentially a problem.
                you can either quit now by pressing ENTER or CTRL+C,
                
                or, if you know what went wrong, you can fix it yourself,
                then manually run the command '/bin/dd if=/dev/zero of=/virtual_0/.fstransform.loop.308145 bs=1 count=1 seek=32635174780927'
                (or something equivalent)
                and finally resume this script by typing CONTINUE and pressing ENTER:
```

**Comment by cosmos72** [comment]
Hi @rswarthout
the error you posted
```
17:45:36 dd: failed to truncate to 32635174780927 bytes in output file '/virtual_0/.fstransform.loop.308145': File too large
```
indicates that the **initial** file system type (ext4) does not support creating the huge sparse file
(32.6 TB, as large as the device itself) needed by fstransform.

Wikipedia (https://en.wikipedia.org/wiki/Ext4#Features) says that maximum file size supported by ext4 is 16 TiB = 17.6 TB,
so it matches your issue.

While this problem is in theory relatively easy to workaround in fstransform (by splitting the huge sparse file into multiple files, and using `dmsetup` instead of `mount -t loop` to create a single block device on top of them), it would require some development.

In summary, there is currently no way to convert an ext4 file system larger than 16 TiB with fstransform. Sorry.

**Comment by rswarthout** [comment]
@cosmos72 Thank you for getting back to me on this. I will proceed with doing a copy of the data to another volume.
